compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -171,37 +171,47 @@ object Inferencing {
     res
   }
 
-  /** This class is mostly based on IsFullyDefinedAccumulator.
-    * It tries to approximate the given type based on the available GADT constraints.
-    */
+  /** Approximates a type to get rid of as many GADT-constrained abstract types as possible. */
   private class ApproximateGadtAccumulator(implicit ctx: Context) extends TypeMap {
 
     var failed = false
 
-    private def instantiate(tvar: TypeVar, fromBelow: Boolean): Type = {
-      val inst = tvar.instantiate(fromBelow)
-      typr.println(i"forced instantiation of ${tvar.origin} = $inst")
-      inst
-    }
-
-    private def instDirection2(sym: Symbol)(implicit ctx: Context): Int = {
-      val constrained = ctx.gadt.fullBounds(sym)
-      val original = sym.info.bounds
-      val cmp = ctx.typeComparer
-      val approxBelow =
-        if (!cmp.isSubTypeWhenFrozen(constrained.lo, original.lo)) 1 else 0
-      val approxAbove =
-        if (!cmp.isSubTypeWhenFrozen(original.hi, constrained.hi)) 1 else 0
-      approxAbove - approxBelow
-    }
-
-    private[this] var toMaximize: Boolean = false
-
+    /** GADT approximation proceeds differently from type variable approximation.
+      *
+      * Essentially, what we're doing is we're inferring a type ascription that
+      * will remove as many GADT-constrained types as possible. This means that
+      * we want to approximate type T to type S in such a way that no matter how
+      * GADT-constrained types are instantiated, T <: S. In other words, the
+      * relationship _necessarily_ must hold.
+      *
+      * We accomplish that by:
+      *   - replacing covariant occurences with upper GADT bound
+      *   - replacing contravariant occurences with lower GADT bound
+      *   - leaving invariant occurences alone
+      *
+      * Examples:
+      *   - If we have GADT cstr A <: Int, then for all A <: Int, Option[A] <: Option[Int].
+      *     Therefore, we can approximate Option[A] ~~ Option[Int].
```

Comment:
I'm still concerned about the stuff I mentioned in https://github.com/lampepfl/dotty/issues/9274#issuecomment-653130226, it seems that if I have a method `def foo[T <: X] = ...`, then I immediately get a GADT constraint `T <: X`, and based on this documentation I would guess this means that ApproximateGadtAccumulator will replacing `Option[T]` by `Option[X]`, even though there's no need to do any approximation and nothing GADT related is actually going on?

Reply:
OK, I now realize that approximateGADT is only called when the type of the tree is not a subtype of the expected type, so it's probably OK technically, but I think it's confusing and not great for performance that we run this stuff in situations which have nothing to do with GADTs. A few suggestions:
- Add some documentation above the call to approximateGADT in adaptToSubType, describing in what situation this might help (e.g., with a simple example)
- As far as I can tell, the result of the approximation is ignored unless `pt` is a `SelectjonProto` and `ctx.gadt.nonEmpty`, so we should only run the type map when these conditions are true, this is better for performance and for understanding what's going on.
- In https://github.com/lampepfl/dotty/issues/9274 you said:
  >  it's not possible to tell whether a symbol is a type parameter of an enclosing function after descending into its body; therefore, we need to stick those symbols somewhere when we see them and we stick them into GadtConstraint
   Type parameters have the `TypeParam` flag set, and I think method type parameters should always be TypeRefs where the prefix is `NoPrefix`, is that enough?

Reply:
Re. documentation: I'll add the doc to code in `adaptToSubType`. We need to approximate GADTs in `adapt` in order to look up members that should be there b/c of GADT constraints. We cannot take GADT constraints into account when doing normal member lookup b/c of cache. Sample code is https://github.com/lampepfl/dotty/pull/9322/files#diff-70e7287c36d5e6a572c97c2b51613049R1-R9.

Re. lazily performing the approximation: will do, doing it eagerly is a remainder from the old `recover`, which also used the result of approx.

Re. type parameters: key issue is telling whether the type parameters are from an enclosing method, or from somewhere else (i.e. a class?) Now that I think about it, there may be some way of doing so. Question is, do we actually get any performance out of it. I'd prefer to know for sure that there's a performance issue with this code before trying to change it.

Final concern: can you take a look @ https://github.com/lampepfl/dotty/pull/9322/files#diff-1b5c8ce1eaca91d793eef217487d5a59R754 and see whether that change makes sense to you? `isNewSubType` would at most return `false` early before doing the type comparison, so I cannot really see any way that this change is wrong, but at the same time all the justification I have for it is that no test broke and it fixes my issue.

Reply:
> Re. type parameters: key issue is telling whether the type parameters are from an enclosing method, or from somewhere else (i.e. a class?)

Check if the owner is a method, type parameters from methods which are not enclosing should not be in scope at all.

> Final concern: can you take a look @ #9322 (files) and see whether that change makes sense to you? 

I guess that usage of isNewSubType was incorrect before? Maybe something worth asking Martin about.

Reply:
> > Re. type parameters: key issue is telling whether the type parameters are from an enclosing method, or from somewhere else (i.e. a class?)
> 
> Check if the owner is a method, type parameters from methods which are not enclosing should not be in scope at all.

Yeah, that's what I was thinking about when I said "there may be some way of doing so". I'd still like an indication that we have a performance problem here.

> > Final concern: can you take a look @ #9322 (files) and see whether that change makes sense to you?
> 
> I guess that usage of isNewSubType was incorrect before? Maybe something worth asking Martin about.

@odersky can you take a look at the single line modified in TypeComparer in https://github.com/lampepfl/dotty/pull/9322/files#diff-1b5c8ce1eaca91d793eef217487d5a59R754 and check if the change looks OK to you? The problem that this change fixes is https://github.com/lampepfl/dotty/pull/9322#issuecomment-655570104.

Reply:
> I'd still like an indication that we have a performance problem here.

Even if it's not a performance problem, I think it's nicer if GADT code doesn't run when nothing GADT related is going on: it makes it easier to know what code can be safely ignored when trying to understand or debug something, and it makes the output of debug printers easier to read (especially the subtyping printer which is hard enough to follow in the best of situations), but this isn't a blocker for this PR, and if you think it would make the resulting GADT logic too complicated then we can keep the current logic of course.